File: Python/Copy-List-with-Random-Pointer.py
# Definition for singly-linked list with a random pointer.
class RandomListNode:
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        d = {}
        dummy = RandomListNode(0)
        p = dummy
        n_head = head
        while n_head != None:
            node = RandomListNode(n_head.label)
            p.next = node
            d[node.label] = node
            p = p.next
            n_head = n_head.next
        p.next = None
        n_head = head
        r_head = dummy.next
        while r_head != None:
            if n_head.random == None:
                r_head.random = None
            else:
                cnt = n_head.random.label
                r_head.random = d[cnt]
            r_head = r_head.next
            n_head = n_head.next
        return dummy.next

File: C++/Copy-List-with-Random-Pointer.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
    // can not modify the original List
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        if (head == NULL) return NULL;
        RandomListNode *pre = new RandomListNode(0), *cur = pre;
        RandomListNode *p = head;
        while (p != NULL) {
            RandomListNode* temp = p->next;
            RandomListNode* node = new RandomListNode(p->label);
            p->next = node;
            node->next = temp;
            p = temp;
        }
        RandomListNode *cnt = head;
        while (cnt != NULL) {
            if (cnt->random != NULL) cnt->next->random = cnt->random->next;
            else cnt->next->random = NULL;
            cnt = cnt->next->next;
        }
        while (head != NULL) {
            RandomListNode* temp = head->next->next;
            cur->next = head->next;
            cur = cur->next;
            head->next = temp;
            head = temp;
        }
        cur->next = NULL;
        return pre->next;
    }
};